core/domain/platform_parameter_domain.py

```diff
@@ -35,3 +63,721 @@ class PlatformParameterChange(change_domain.BaseChange):
+
+    SUPPORTED_FILTER_TYPES = [
+        'server_mode', 'user_locale', 'client_type', 'browser_type',
+        'app_version',
+    ]
+
+    SUPPORTED_OP_FOR_FILTERS = {
+        'server_mode': ['='],
+        'user_locale': ['='],
+        'client_type': ['='],
+        'browser_type': ['='],
+        'app_version': ['=', '<', '<=', '>', '>='],
+    }
+
+    def __init__(self, filter_type, conditions):
+        self._type = filter_type
+        self._conditions = conditions
+
+    @property
+    def type(self):
+        """Returns filter type.
+
+        Returns:
+            str. The filter type.
+        """
+        return self._type
+
+    @property
+    def conditions(self):
+        """Returns filter conditions.
+
+        Returns:
+            list((str, str)). The filter conditions. Each element of the list
+            is a 2-tuple (op, value), where op is the operator for comparison,
+            value is the value used for comparison.
+        """
+        return self._conditions
+
+    def evaluate(self, context):
+        """Tries to match the given context with the filter against its
+        value(s).
+
+        Args:
+            context: EvaluationContext. The context for evaluation.
+
+        Returns:
+            bool. True if the filter is matched.
+        """
+        for op, value in self._conditions:
+            is_matched = self._evaluate_single_value(op, value, context)
+            if is_matched:
+                return True
+        return False
+
+    def _evaluate_single_value(self, op, value, context):
+        """Tries to match the given context with the filter against the
+        given value.
+
+        Args:
+            op: str. The operator for comparison, e.g. '='.
+            value: str. The value to match against.
+            context: EvaluationContext. The context for evaluation.
+
+        Returns:
+            bool. True if the filter is matched.
+        """
+        if op not in self.SUPPORTED_OP_FOR_FILTERS[self._type]:
+            raise Exception(
+                'Unsupported comparison operator \'%s\' for %s filter, '
+                'expected one of %s.' % (
+                    op, self._type, self.SUPPORTED_OP_FOR_FILTERS[self._type]))
+        matched = False
+        if self._type == 'server_mode':
+            if op == '=':
+                matched = context.server_mode == value
+        elif self._type == 'user_locale':
+            if op == '=':
+                matched = context.user_locale == value
+        elif self._type == 'client_type':
+            if op == '=':
+                matched = context.client_type == value
+        elif self._type == 'browser_type':
+            if op == '=':
+                matched = context.browser_type == value
+        elif self._type == 'app_version':
+            if context.app_version is not None:
+                matched = self._match_version_expression(
+                    op, value, context.app_version)
```

Comment:
The ifs can be simplified to `if self._type == 'server_mode' and op == '=':`. Also, why do we check `None` only for `app_version` what in the other cases? Should they also return `False` whenever they are `None`?

Reply:
Yes, `False` should be returned for `None` values -- added checks for `None` at the top of the method, simplified to `if self._type == 'server_mode' and op == '=':`

Reply:
Update: sorry I missed something here. `value` must not be `None` and we already have validation for that, while fields of `context` might be `None` (`app_version` and `browser_type`).
For filter other than `app_version`, if the corresponding context field is `None`, the eq comparison is `False` by nature (which is intuitive), but we still need to check if `app_version` is `None` since we are doing regex match -- I have moved the `None` check into `_match_version_expression` so that in `_evaluate_single_value` it looks more consistent.